articles/semi_structured_data.rst

```diff
@@ -0,0 +1,151 @@
+performance. Let’s take a look at what an example schema might look like
+following by a few queries to show what’s possible:
+
+.. code:: sql
+
+    CREATE TABLE visits AS (
+      id UUID,
+      site_id uuid,
+      visited_at TIMESTAMPTZ,
+      session_id UUID,
+      page TEXT,
+      url_params JSONB
+    )
+
+Note that url parameters for an event are open-ended, and no parameters
+are guaranteed. Even the common "utm" parameters (such as utm\_source,
+utm\_medium, utm\_campaign) are by no means universal. Our choice of
+using a
+`JSONB <https://www.citusdata.com/blog/2016/07/14/choosing-nosql-hstore-json-jsonb/>`__
+column for url\_params is much more convenient than creating columns for
+each parameter. With JSONB we can get both the flexibility of schema,
+and combined with `GIN
+indexing <https://www.postgresql.org/docs/9.5/static/gin.html>`__ we can
+still have performant queries against all keys and values without having
+to index them individually.
+
+Enter Citus
+~~~~~~~~~~~
+
+Assuming you do need to scale beyond a single node,
+`Citus <https://www.citusdata.com/product/>`__ can help at scaling out
+your processing power, memory, and storage. In the early stages of
+utilizing Citus you’ll create your schema, then tell the system how you
+wish to shard your data.
+
+In order to determine the ideal sharding key you need to examine the
+query load and types of operations you’re looking to perform. If you are
+storing aggregated data and all of your queries are per customer then a
+shard key such as customer\_id or tenant\_id can be a great choice. Even
+if you have minutely rollups and then need to report on a daily basis
+this can work well. This allows you to easily route queries to shards
+just for that customer. As a result of routing queries to a single shard
+this can allow you a higher concurrency.
+
+In the case where you are storing raw data, there often ends up being a
+lot of data per customer. Here it can be more difficult to get
+sub-second response without further parallelizing queries per customer.
+*It may also be difficult to get predictable sub-second responsiveness
+if you have a low number of customers or if 80% of your data comes from
+one customer.* In the above mentioned cases, picking a shard key that's
+more granular than customer or tenant id can be ideal.
+
+The distribution of your data and query workload is what will heavily
+determine which key is right for you.
+
+With the above example if all of your sites have the same amount of
+traffic then ``site_id`` might be reasonable, but if either of the above
+cases is true then something like ``session_id`` could be a more ideal
+distribution key.
+
+The query workload
+~~~~~~~~~~~~~~~~~~
+
+With a sharding key of ``session_id`` we could easily perform a number
+of queries such as:
+
+Top page views over the last 7 days for a given site:
+
+.. code:: sql
+
+    SELECT page, 
+           count(*)
+    FROM visits
+    WHERE site_id = ‘foo’
```

Comment:
The quotes in this code block (and the one below) look incorrect.

Could we change ‘foo’ with 'foo'?